lib/astutils.cpp

```diff
@@ -3664,3 +3664,63 @@ bool isUnevaluated(const Token *tok)
 {
     return Token::Match(tok, "alignof|_Alignof|_alignof|__alignof|__alignof__|decltype|offsetof|sizeof|typeid|typeof|__typeof__ (");
 }
+
+static std::set<MathLib::bigint> getSwitchValues(const Token *startbrace, bool &hasDefault)
+{
+    std::set<MathLib::bigint> values;
+    const Token *endbrace = startbrace->link();
+    if (!endbrace)
+        return values;
+
+    hasDefault = false;
+    for (const Token *tok = startbrace->next(); tok && tok != endbrace; tok = tok->next()) {
+        if (Token::simpleMatch(tok, "{") && tok->scope()->type == Scope::ScopeType::eSwitch) {
+            tok = tok->link();
+            continue;
+        }
+        if (Token::simpleMatch(tok, "default")) {
+            hasDefault = true;
+            break;
+        }
+        if (Token::simpleMatch(tok, "case")) {
+            const Token *valueTok = tok->astOperand1();
+            if (valueTok->hasKnownIntValue())
+                values.insert(valueTok->getKnownIntValue());
+            continue;
+        }
+    }
+
+    return values;
+}
+
+bool isExhaustiveSwitch(const Token *startbrace)
+{
+    if (!startbrace || !Token::simpleMatch(startbrace->previous(), ") {") || startbrace->scope()->type != Scope::ScopeType::eSwitch)
+        return false;
+    const Token *rpar = startbrace->previous();
+    const Token *lpar = rpar->link();
+
+    const Token *condition = lpar->astOperand2();
+    if (!condition->valueType())
+        return false;
+
+    bool hasDefault{};
+    const std::set<MathLib::bigint> switchValues = getSwitchValues(startbrace, hasDefault);
+
+    if (hasDefault)
+        return true;
+
+    if (condition->valueType()->type == ValueType::Type::BOOL)
+        return switchValues.count(0) && switchValues.count(1);
+
+    if (condition->valueType()->isEnum()) {
+        const Scope *typeScope = condition->valueType()->typeScope;
+        const std::vector<Enumerator> &enumList = typeScope->enumeratorList;
+        const std::size_t usedValues = std::count_if(enumList.begin(), enumList.end(), [&](const Enumerator &e) {
```

Comment:
replace `count_if` with `all_of`. and well I think cbegin/cend feels better than begin/end.

Reply:
Hm, I'm not sure `all_of` would yield the same result. If the user adds cases for values that don't correspond to some enum variants it will fail, even if all enum variants are covered.

Reply:
>  If the user adds cases for values that don't correspond to some enum variants it will fail, even if all enum variants are covered.

I suggest that you just replace `std::count_if` with `std::all_of`. loop through the enum constants, not the switch cases.

there can also be 2 enum constants with the same value and then 1 switch case is enough to cover both:

```
    enum AB {A=1, B=1};
    // .. a switch that has either a "case A:" or a "case B:" is exhaustive..
```

Reply:
Ah, yes, that's better.